# Data Analytics Program
## Python Fundamentals for Data Analysts
## 4.5: Data Consistency Checks
## Alphy Jose

# Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing datasets

In [2]:
# Craeting string of project folder path
path = r'C:\Users\augus\Documents\01-2026 Instacart Basket Analysis' 

In [3]:
# Importing products.csv data set
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [57]:
#Importing  the orders_wrangled Dataset
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Data Consistency Checks

## Mixed-type Data

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\augus\AppData\Local\Temp\ipykernel_27596\2671172800.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [11]:
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [12]:
# Finding missing values in products dataset
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
#Creating a subset dataframe with missing entries only
df_nan = df_prods[df_prods['product_name'].isnull() == True ]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_prods.shape

(49693, 5)

In [16]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [19]:
# Creating new subset dataframe with duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
# creating dataframes without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task

## Step 2 :Run the df.describe() function on your df_ords dataframe

In [58]:
# Running describe function on the df_ords dataframe
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### As per my initial analysis, all values in cloumns falls within acceptable range. For orders_day_week , the min value is 0(saturday) and max value is 6(friday). Similarly for order_hour_of_day,the min value is 0 and max value is 23. All columns have same count of values except the days_since_prior_order 

In [66]:
# Running describe function on the df_prds dataframe
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### The only issue I can find from these stats is the maximum value of prices column, which is 99999 and it is not acceptable compared to other values in prices column.

In [67]:
#Lookinf for outliers in the df_prods dataframe.
df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] > df_prods_clean_no_dups['prices'].mean() + (2 * df_prods_clean_no_dups['prices'].std())]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [68]:
#Fidning the median of the df_prods dataframe
df_prods_clean_no_dups['prices'].median() 

7.1

In [71]:
#Replacing  the outliers values in the prices column with the meadian value of 7.1.
df_prods_clean_no_dups = df_prods_clean_no_dups.replace({"prices":{14900.0:7.1, 99999.0:7.1}})

In [70]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      7.680414
std    14340.705287     38.315784       5.850779      4.199298
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

## Step 3: Check for mixed-type data in your df_ords dataframe.

In [25]:
# Checking for mixed-type columns
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\augus\AppData\Local\Temp\ipykernel_27596\210590000.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\augus\AppData\Local\Temp\ipykernel_27596\210590000.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\augus\AppData\Local\Temp\ipykernel_27596\210590000.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\augus\AppData\Local\Temp\ipykernel_27596\210590000.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\augus\AppData\Local\Tem

#### There are no columns with mixed data types in orders dataframe

## Step 5:Run a check for missing values in your df_ords dataframe

In [59]:
# checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### There are 206209 missing values in the column days_since_prior_order.This column represents the number of days that has passed after the previous order.So for the first order of a new user, the value for this column could be missing. So I assume it is not actually a missing value.

## Step 6:Address the missing values using an appropriate method

In [61]:
#Creating a subset dataframe with missing entries only
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True ]

In [62]:
df_ords_nan.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
11   2168274        2             1                   2                 11   
26   1374495        3             1                   1                 14   
39   3343014        4             1                   6                 11   
45   2717275        5             1                   3                 12   

    days_since_prior_order  
0                      NaN  
11                     NaN  
26                     NaN  
39                     NaN  
45                     NaN

In [63]:
df_ords_nan.describe()

order_id        user_id  order_number  orders_day_of_week  \
count  2.062090e+05  206209.000000      206209.0       206209.000000   
mean   1.708462e+06  103105.000000           1.0            2.754118   
std    9.881299e+05   59527.555167           0.0            2.076205   
min    2.000000e+01       1.000000           1.0            0.000000   
25%    8.507300e+05   51553.000000           1.0            1.000000   
50%    1.706246e+06  103105.000000           1.0            3.000000   
75%    2.564292e+06  154657.000000           1.0            5.000000   
max    3.421081e+06  206209.000000           1.0            6.000000   

       order_hour_of_day  days_since_prior_order  
count      206209.000000                     0.0  
mean           13.626597                     NaN  
std             4.223769                     NaN  
min             0.000000                     NaN  
25%            11.000000                     NaN  
50%            14.000000                     NaN  
75%            17.000000                     NaN  
max            23.000000                     NaN

#### From the statistics, we can see that the order number is always 1.

In [32]:
# checking whether user_id's are repeated or not
count_of_duplicates = df_ords_nan['user_id'].duplicated().sum()

In [33]:
count_of_duplicates

np.int64(0)

#### All user id's are unique, so we can conclude that missing values in 'days_since_prior_order' are the first orders of these 206209 users. And there is no need to remove these missing values.

## Step 7:Run a check for duplicate values in your df_ords data

In [34]:
# checking for duplicates
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### There are no duplicate in orders dataset

## Step 9:Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder

In [72]:
# Exporting products dataset.
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'),index=False)

In [64]:
# Exporting orders dataset.
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'),index=False)